### PR TITLE
Add intermediateDecodeExpensive() function and Python binding

### LINKED ISSUE #1953: Bug: intermediateDecode ignores unprocessed internal buffer
**Describe the bug**
Stream.intermediateDecode() ignores part of the buffer data that has been sent to the stream.

**To Reproduce**
(In Python)
```
# 1. Create a stream
model = stt.Model(args.model)
stream_context = model.createStream()

# 2. Feed it a buffer
stream_context.feedAudioContent(some_buffer)

# 3. Do an intermediate decode
text = stream_context.intermediateDecode()
```

**Current behavior**
Depending on the size of some_buffer:
- If some_buffer is smaller than the internal batch size, nothing will be processed.
- If some_buffer is larger than the internal batch size, only the first part will be processed.

**Expected behavior**
All of the audio data that has been sent to the stream API so far would be processed and an intermediate result returned.

This is important when streaming to STT with VAD, where a stream will be automatically stopped whenever VAD detection is negative. If the stream is stopped at a point that is not an exact multiple of the internal batch buffer, parts of the audio will not have been processed by the acoustic model and therefore will be missing from the intermediate result. This causes mis-recognition of words.

Stream.finalizeStream() does not suffer from this defect, but it cannot provide intermediate results because it shuts down the stream completely. Intermediate results are essential for fast / low latency voice recognition.

**Environment (please complete the following information):**
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Custom Yocto Linux
- **TensorFlow installed from (our builds, or upstream TensorFlow)**: Coqui Tensorflow
- **TensorFlow version (use command below)**:
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: 3.1.0
- **GCC/Compiler version (if compiling from source)**: Yocto cross compiler
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A

### PR DESCRIPTION
**Pull request by jeremiahrose**
This flushes the buffer through the acoustic model before returning
an intermediate decode. Provides an alternative to finishStream()
that doesn't end the stream.

Fixes #1953

I added the Python bindings manually, as I haven't used SWIG before. Let me know there's any glaring mistakes. I haven't added any other bindings so if you have any pointers on how to do that I will happily oblige.

Let me know if there's anything else that needs changing!

The function seems to be working very well on our test hardware and is improving command response with the unreliable webrtc VAD a great deal!

Built and tested on v0.10.0-alpha.9 because I couldn't get later versions to compile.

### CODE REVIEW

File: native_client/stt.cc [hunk 1]
```diff
@@ -167,6 +176,23 @@ StreamingState::processAudioWindow(const vector<float>& buf)
   pushMfccBuffer(mfcc);
 }
 
+void
+StreamingState::flushBuffer()
+{
+  // Flush audio buffer
+  processAudioWindow(audio_buffer_);
+
+  // Add empty mfcc vectors at end of sample
```
Comment (JRMeyer) [inline native_client/stt.cc:185]: Is this a comment, or commented-out code? Why?

File: native_client/stt.cc [hunk 2]
```diff
@@ -167,6 +176,23 @@ StreamingState::processAudioWindow(const vector<float>& buf)
...
+  //for (int i = 0; i < model_->n_context_; ++i) {
+  //  addZeroMfccWindow();
+  //}
```
Comment (JRMeyer) [inline native_client/stt.cc:188]: Is this a comment or commented-out code? Why is this needed?

Comment (jeremiahrose) [reply native_client/stt.cc]: Oops. Fixed

File: native_client/coqui-stt.h [hunk 3]
```diff
@@ -326,6 +326,20 @@ Metadata* STT_IntermediateDecodeWithMetadata(const StreamingState* aSctx,
  * @note This method will free the state pointer (@p aSctx).
  */
 STT_EXPORT
+char* STT_IntermediateDecodeExpensive(StreamingState* aSctx);
+
+/**
+ * @brief Compute the intermediate decoding of an ongoing streaming inference,
```
Comment (JRMeyer) [inline native_client/coqui-stt.h:332]: This `@brief` will go in the docs, so can we expand it more, like you did for the python bindings?

Comment (JRMeyer) [reply native_client/coqui-stt.h:332]: Also, explain in a few words why this is "expensive"

Comment (jeremiahrose) [reply native_client/coqui-stt.h]: Roger that

File: native_client/stt.cc [hunk 4]
```diff
@@ -167,6 +176,18 @@ StreamingState::processAudioWindow(const vector<float>& buf)
...
+{
+  // Flush audio buffer
+  processAudioWindow(audio_buffer_);
+
+  // Process final batch
+  if (batch_buffer_.size() > 0) {
+    processBatch(batch_buffer_, batch_buffer_.size()/model_->mfcc_feats_per_timestep_);
+  }
+}
+
```
Comment (reuben) [inline native_client/stt.cc:190]: Maybe rework `finalizeStream` instead of creating a new method? You can rename it to `flushBuffers` and then add a `addZeroMfccVectors=false` parameter which is then set to true when called by `finishStream` and `finishStreamWithMetadata`.

File: native_client/coqui-stt.h [hunk 5]
```diff
@@ -326,6 +326,23 @@ Metadata* STT_IntermediateDecodeWithMetadata(const StreamingState* aSctx,
...
+ *
+ * @param aSctx A streaming state pointer returned by {@link STT_CreateStream()}.
+ * @param aNumResults The number of candidate transcripts to return.
+ *
+ * @return Metadata struct containing multiple candidate transcripts. Each transcript
+ *         has per-token metadata including timing information. The user is
+ *         responsible for freeing Metadata by calling {@link STT_FreeMetadata()}.
+ *         Returns NULL on error.
+ */
+STT_EXPORT
```
Comment (reuben) [inline native_client/coqui-stt.h:345]: This comment is on the wrong function :)

**Review by reuben — COMMENTED** [review]
Thanks for the PR! This needs a bit more work before it's merged, but the hardest part you have already figured out. Our *Decode functions have a corresponding *DecodeWithMetadata` version. Can you add a corresponding `STT_IntermediateDecodeExpensiveWithMetadata` which returns full metadata results?

Adding support for the JavaScript binding should be simple and basically the same work as you did for the Python bindings. Add the new method and doc comment to the `Stream` class in `native_client/javascript/index.ts`, then add a `%newobject %newobject STT_IntermediateDecodeExpensive;` line to `native_client/javascript/stt.i`, then rebuild and check either manually on a REPL or by modifying `native_client/javascript/client.ts`.

It would also be great if you could add a way to exercise this function in `native_client/client.cc`. See for example the `stream_size > 0` and `extended_stream_size > 0` checks in `LocalDsSTT` in `native_client/client.cc`. Those are used in our CI to exercise these API paths.

**Comment by jeremiahrose** [comment]
I had some issues testing the js bindings. I'm on alpha 9 so you might want to check it against a newer version of STT. The error I was getting was 
```
../stt_wrap.cxx:2141:34: error: macro "SWIGV8_ARRAY_NEW" passed 1 arguments, but takes just 0
     jsresult = SWIGV8_ARRAY_NEW(0);
```

File: native_client/coqui-stt.h [hunk 6]
```diff
@@ -326,6 +361,20 @@ Metadata* STT_IntermediateDecodeWithMetadata(const StreamingState* aSctx,
...
```
Comment (reuben) [inline native_client/coqui-stt.h:377]: Looks like somehow the merges reintroduced the problem of this comment being on the wrong function?

**Review by reuben — COMMENTED** [review]
Looks like some git-related issues popped up, but otherwise just a minor comment on the JS binding and then some code to exercise these new methods in tests and this will be good to merge!

**Comment by reuben** [comment]
Also the build should be fixed on main, if you're still using Dockerfile.build

**Comment by reuben** [comment]
(FWIW, you probably /shouldn't/ use Dockerfile.build for development, as Docker's caching is not as good as bazel's, so you'll waste a lot of cycles).

**Comment by jeremiahrose** [comment]
Sorry for the delay! Fixed the minor errors you listed above, and merged in the latest main. Not sure how you want the intermediateDecodeExpensive js bindings to be tested. I had a look at asserts.sh and there only seem to be tests for the stt binary?
Might not have time to develop the tests.

**Comment by reuben** [comment]
Sorry for dropping the ball here. I'll rebase and take a look at what's missing. The `stt` binary is exposed from the JS bindings (`native_client/javascript/client.ts`) too.

**Comment by reuben** [comment]
Also I decided to rename the methods to `FlushBuffers` instead of `Expensive` (better to name for behavior than for side-effect, and the side-effect is in the docs).
